### PR TITLE
VCU118 FPGA Updates + FireMarshal on Prototypes

### PR DESCRIPTION
**Pull request by abejgonzalez**
**Related issue**: https://github.com/ucb-bar/chipyard/issues/822 https://github.com/firesim/FireMarshal/pull/190

**Type of change**: other enhancement

**Impact**: software change + other

**Release Notes**
Enlarge the payload from 20MiB to 30MiB for the SDCard boot (for larger FireMarshal images). Use `DefaultClockFrequencyKey` to determine the FPGA frequency instead of another key. Add documentation for using FireMarshal for the prototype platform.

TODO:
- [x] Write https://github.com/ucb-bar/chipyard/issues/822 into actual documentation
- [x] Bump FireMarshal for prototype SDCard loading (https://github.com/firesim/FireMarshal/pull/190)

### CODE REVIEW

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala
Comment (abejgonzalez) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:48]: Why was this necessary? Why not just use the PBUS frequency specified?

Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:48]: We want to move towards explicitly specifying all frequencies. It's not obvious that pbus freq sets mbus freq when mbus freq is unspecified.

Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:48]: In fact the behavior of pbus freq setting mbus freq is the result of a workaround on our side.

Comment (abejgonzalez) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:48]: This prevents the `DefaultClockFrequencyKey` being used for the MBUS... requiring an extra config. fragment to override  it (either a fragment to set the MBUS freq to the PBUS freq if the PBUS frequency changes or a fragment to reset the MBUS frequency to the `DefaultClockFrequencyKey`). I think we can just modify the comment stating that in this default setup the MBUS == PBUS frequency.

Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:48]: The DefaultClockFrequencyKey was the workaround. In an ideal world, all clock sinks would explicitly set their clock frequencies.

If some config inheriting this wants to change the default mbus/pbus frequencies, it should do so explicitly, through using two config fragments, rather than relying on knowledge of default clock setting behaviors.

@davidbiancolin

Comment (abejgonzalez) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:48]: This also seems to break things that check the DTS frequency (since its an `Option`)... so anything that checks a DTS frequency would also have to check the `DefaultClockFrequencyKey` if the `Option` returns `None`.

Comment (abejgonzalez) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala]: I reverted this.

**Comment by abejgonzalez** [comment]
Note: The doc build passes locally. The doc error seems to happen on other PRs so this isn't an issue with these docs.

**Comment by colinschmidt** [comment]
CI only started failing the doc check after this PR: https://github.com/ucb-bar/chipyard/pull/846
It seems to fail on `dev` specifically but not on the PR branches

**Comment by jerryz123** [comment]
Chisel changes LGTM. Someone who has done bringup should review the doc changes.

**Review by jerryz123 — APPROVED** [review]

**Comment by abejgonzalez** [comment]
Ping @alonamid @colinschmidt .

**Comment by alonamid** [comment]
Could you please add commas in the docs page?

**Comment by abejgonzalez** [comment]
I won't re-run CI unless necessary since it passed before the grammar changes.

File: docs/Prototyping/VCU118.rst
Comment (colinschmidt) [inline docs/Prototyping/VCU118.rst:108]: "to store a file system that can be accessed from the DUT".

File: fpga/src/main/resources/vcu118/sdboot/sd.c
Comment (colinschmidt) [inline fpga/src/main/resources/vcu118/sdboot/sd.c:12]: How much extra room is there here?

File: fpga/src/main/resources/vcu118/sdboot/sd.c
Comment (colinschmidt) [inline fpga/src/main/resources/vcu118/sdboot/sd.c:175]: Why do we need both of these messages? The second loading seems redundant.

File: fpga/src/main/resources/vcu118/sdboot/sd.c
Comment (colinschmidt) [inline fpga/src/main/resources/vcu118/sdboot/sd.c:177]: I would either file an issue to look at this in the future or just delete this comment and the other line.

**Review by colinschmidt — APPROVED** [review]
LGTM. Just some minor docs changes

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst:82]: This file doesn't exist unless the user makes it. So this is assuming we create it from scratch.

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst:72]: I assume it probably can, but I don't know how to do that (nor do I think it will make much of a difference). Having it as a block device would change things in a couple ways:

1. Have to figure out how to get it to work with Linux / SPI interface. Create new partitions? Have a block device driver? Etc.
2. Could probably minimize the binary size. Thus making loading faster if we change the bootrom.

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst:62]: The overall section splits the VCU118 into 2 sections the "Basic Design" and "Platform", I can modify this a bit to be "Basic Design" and "Bringup Design" and change this title to be clearer.

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst:88]: It does nothing special with the UART stuff... so its the default UART driver.

File: fpga/src/main/resources/vcu118/sdboot/sd.c
Comment (abejgonzalez) [reply fpga/src/main/resources/vcu118/sdboot/sd.c:12]: The Linux binary with FireMarshal is around 20-25MiB so 5MiB room.

File: fpga/src/main/resources/vcu118/sdboot/sd.c
Comment (abejgonzalez) [reply fpga/src/main/resources/vcu118/sdboot/sd.c:175]: There is a nice spinner that comes up next to the loading so I just left this. Technically I can put the spinner next to the `kprintf` that I added, but I figured this extra print wasn't a bottleneck so just leave it.

File: fpga/src/main/resources/vcu118/sdboot/sd.c
Comment (abejgonzalez) [reply fpga/src/main/resources/vcu118/sdboot/sd.c:177]: https://github.com/ucb-bar/chipyard/issues/854

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:87]:
```suggestion
3. Next, build the workload (a.k.a buildroot Linux) in FireMarshal with the ``nodisk`` option flag.
```

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:95]:
```suggestion
.. Note:: Using the "board" FireMarshal functionality allows any child workload depending on the ``br-base.json`` workload specification to target a "prototype" platform rather than FireChip platform.
```

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:99]: is this true? wouldn't the image and binary be created in ``$PATH_TO_FIREMARSHAL/images`` right after the ``build`` command? why do you need to install?

Comment (alonamid) [reply docs/Prototyping/VCU118.rst:99]: oh, does the install command at the ``-flat``?

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:110]:
```suggestion
Additionally, these instructions assume you are using Linux with ``sudo`` privileges and ``gdisk``, but you can follow a similar set of steps on Mac (using ``gpt`` or another similar program).
```

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:120]: would it be possible to give prompt examples? (..code-block:: shell ....)

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:129]:
```suggestion
   For the type, search for the `apfs` type and use the hex number given.
```

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:133]:
```suggestion
   For the type, search for the `hfs` and use the hex number given.
```

File: docs/Prototyping/VCU118.rst
Comment (alonamid) [inline docs/Prototyping/VCU118.rst:165]:
```suggestion
Once connected, you should see the binary being loaded as well as Linux output (in some cases you might need to reset the DUT).
```

**Review by alonamid — APPROVED** [review]
LGTM pending commas

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst:99]: Correct, the binary/image are created at 1st (technically just the binary with `nodisk`), then the `install` flattens the binary so it can be `dd`'ed onto the SDCard

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst:120]: I think that is a bit excessive and unnecessary. These instructions although are setup for `gdisk` are also a bit abstract to allow others to port to other disk programs. If people get stuck on this step we can add more details if need be (I presume this will be the easiest step with the FireMarshal setup being the trickiest).

File: docs/Prototyping/VCU118.rst
Comment (abejgonzalez) [reply docs/Prototyping/VCU118.rst]: Added a bit of clarification.

**Comment by jerryhethatday** [comment]
@abejgonzalez Hello, I want to run my `rv32imafc `chipyard configuration on vcu118, and test it with baremetal software. Since chipyard currently does not support boot from flash, I therefore want to compile a simple rv32 helloworld program, create `helloword.bin` using objcopy, and put it in specific sectors of sdcard following instructions in chipyard doc. My confusion is that do I need FireMarshal  to generate the baremetal workload? I read somewhere in FireMarshal that it also supports build baremetal workloads. I wonder if my method will work. And Is there some extra considerations for this? Thanks!

**Pull request merged**
